### PR TITLE
feat(hydroflow_plus, stageleft)!: allow cluster self ID to be referenced as a global constant

### PR DESCRIPTION
**Pull request by shadaj**
This eliminates the need to store `cluster.self_id()` in a local variable first, instead you can directly reference `CLUSTER_SELF_ID`.
---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/hydro-project/hydroflow/pull/1574).
* #1583
* #1575
* __->__ #1574

### CODE REVIEW

**Comment by shadaj** [comment]
Guide for review. The first big change is to stageleft, where we create a richer version of `Quoted` that takes a `Ctx` type parameter. The idea is that `Ctx` captures some metadata about _where_ the quoted snippet will be spliced, which can be used to augment how free variables are made available inside the quoted snippet.

In Hydroflow+, the context for UDFs is the location for the stream, so this allows us to have a general global `CLUSTER_SELF_ID` value which when used as a free variable will return the appropriate `ClusterId` based on the provided context. This preserves the type safety guarantees, but avoids needing to call `cluster.self_id()`. In fact, this API was previously unsafe, since you could call `c1.self_id()` and reference that in a UDF on `c2`, which would result in malformed generated code. Now, because we rely on the context, we know that the spliced free variable will always be well-formed.

File: stageleft/src/runtime_support.rs [hunk 1]
```diff
@@ -75,11 +75,29 @@ impl ParseFromLiteral for bool {
 impl_parse_from_literal_numeric!(i8, i16, i32, i64, i128, isize);
 impl_parse_from_literal_numeric!(u8, u16, u32, u64, u128, usize);
 
-pub trait FreeVariable<O> {
-    fn to_tokens(self) -> (Option<TokenStream>, Option<TokenStream>)
+pub trait FreeVariableWithContext<Ctx> {
```
Comment (MingweiSamuel) [inline stageleft/src/runtime_support.rs:78]: (Maybe default `Ctx=()`?)

**Review by MingweiSamuel — APPROVED** [review]
Is `CLUSTER_SELF_ID` just a magic name? In the future probably best to namespace it to something, `ctx.CLUSTER_SELF_ID`, if/when we add other magic names

**Comment by shadaj** [comment]
> Is `CLUSTER_SELF_ID` just a magic name? In the future probably best to namespace it to something, `ctx.CLUSTER_SELF_ID`, if/when we add other magic names

Ah no, it's actually defined as a `const CLUSTER_SELF_ID` in the `cluster` module!

File: stageleft/src/runtime_support.rs (hunk 1 above)
Comment (shadaj) [reply stageleft/src/runtime_support.rs:78]: We want free variables to be as generic wrt contexts as possible, so this forces developers to do that.

**Pull request merged**